exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex

```diff
@@ -0,0 +1,60 @@
+defmodule DancingDots.DotGroup do
+  # This module is an example of how behaviours can be used in practice.
+  # You don't need to read it to solve this exercise.
+  # It's here for the curious :)
```

Comment:
I'm not sure if it's a good idea to leave this code here. But one of my biggest problems with understanding behaviours is actually imagining how they can be used once I have some modules that all implement that one behaviour.

Reply:
Yeah, I'm with you, it's nice to see how it can be used. `render_dots/2` is great because it's very readable, but I think maybe `new/2` is a bit too complex... Maybe it could be changed to filter the valid options rather than throwing an error for the sake of simplicity?

(untested idea)
```elixir
def new(dots, animations_with_opts) do
  valid_animations_with_opts =
  for {animation_module, opts} <- animations_with_opts,
      # using Animation's init/1 callback
      {:ok, opts} <- animation_module.init(opts) do
    {animation_module, opts}
  end

  {:ok,  %__MODULE__{dots: dots, animations_with_opts: valid_animations_with_opts }}
end

```

Reply:
I agree, `new` was too complex. I think the worst offender there was the `reduce_while` call. To make it simpler, I decided to split `new` into two functions - `new` that just returns a group with no animations, and `add_animation` to add a single animation module to the list. That allows getting rid of any `Enum` calls.

I didn't want to silently ignore initialization errors in this demo because that would show an anti-pattern IMO.

Reply:
Yes, this new version is great!